hazelcast/src/test/java/com/hazelcast/client/alto/AltoChannelConnectorTest.java

```diff
@@ -0,0 +1,254 @@
+    @Test
+    public void testConnector_whenConnectionIsClosed_afterChannelsAreSet() throws IOException {
+        doAnswer(invocation -> {
+            // While calling the setAltoChannels, simulate a closed
+            // connection by returning false in the isAlive method
+            when(mockConnection.isAlive()).thenReturn(false);
+            return null;
+        }).when(mockConnection).setAltoChannels(any());
+
+        connector.initiate();
+
+        // We should set the Alto channels
+        verify(mockConnection, times(1)).setAltoChannels(any());
+        assertFalse(mockConnection.isAlive());
+
+        // But, the channels should be closed, as the connection is
+        // no longer alive
+        for (Channel channel : mockAltoChannels) {
+            verify(channel, times(1)).write(any());
+            verify(channel, times(1)).close();
+        }
+    }
+
+    @Test
+    public void testConnector_whenConnectionIsClosed_afterSomeChannelsAreEstablished() throws IOException {
+        OngoingStubbing<Channel> stubbing = when(mockChannelCreator.create(any(), any(), any()));
+        int count = 0;
+        for (Channel channel : mockAltoChannels) {
+            if (++count < mockAltoChannels.length) {
+                // Return the channel successfully for any channel creation
+                // other than the last one
+                stubbing = stubbing.thenReturn(channel);
+            }
+        }
+
+        // Simulate connection failure while creating the last channel
+        stubbing.thenAnswer(invocation -> {
+            when(mockConnection.isAlive()).thenReturn(false);
+            return mockAltoChannels[mockAltoChannels.length - 1];
+        });
+
+        connector.initiate();
+
+        verify(mockConnection, never()).setAltoChannels(any());
+        assertFalse(mockConnection.isAlive());
+
+        for (Channel channel : mockAltoChannels) {
+            verify(channel, times(1)).write(any());
+            verify(channel, times(1)).close();
+        }
+    }
+
+    @Test
+    public void testConnector_whenChannelCreationsFails_afterSomeChannelsAreEstablished() throws IOException {
+        OngoingStubbing<Channel> stubbing = when(mockChannelCreator.create(any(), any(), any()));
+        int count = 0;
+        for (Channel channel : mockAltoChannels) {
+            if (++count < mockAltoChannels.length) {
+                // Return the channel successfully for any channel creation
+                // other than the last one
+                stubbing = stubbing.thenReturn(channel);
+            }
+        }
+
+        // Throw while creating the last channel
+        stubbing.thenThrow(new RuntimeException("expected"));
+
+        connector.initiate();
+
+        verify(mockConnection, never()).setAltoChannels(any());
+        for (Channel channel : mockAltoChannels) {
+            if (--count == 0) {
+                // The last channel is not even "created",
+                // we have thrown exception instead.
```

Comment:
Can we somehow assert that the last channel in the channel connector's variable is null? We may also assert that none of the last channel's methods is called. (we have tha channel in mockAltoChannels, it is just not added to alto channel connector object's array)

Reply:
The first one seems like an implementation detail. I have added checks that make sure none of the methods are called, but not doing the first one, as it requires me to make the private field visible to the test, which I don't like